pkg/kubecost/asset.go

```diff
@@ -2614,12 +2662,10 @@ func (as *AssetSet) accumulate(that *AssetSet) (*AssetSet, error) {
 	}
 
 	// In the case of an AssetSetRange with empty entries, we may end up with
-	// an incoming as without props, even though we are trying to aggregate
-	// by props. This handles that case, assigning the correct props.
-	if !propsEqual(as.props, that.props) {
```

Comment:
I think we need to keep this call to `!propsEqual`, and perhaps I should make the comment more clear. Basically, during accumulation, it's possible to accumulate over a range where the first `as` is empty (and thus has empty `as.props`), but at some point we hit "real" data that has non-empty `props`; e.g. accumulating some 5d range where the JSON would look like:
```json
[
  {},
  {},
  {
    "default": {...},
    "kubecost": {...}
  },
  {
    "default": {...},
    "kubecost": {...}
  },
  {
    "default": {...},
    "kubecost": {...}
  },
]
```
The first two sets don't have props, but the last three have `props == ["namespace"]`. So we want to persist those props, not the empty ones.

Does that make sense? I get some funky code smell from this, so we may just have to revisit with a less error-prone design at some point.

Reply:
I will make a note to go look in the unit tests to make sure we have coverage for this case. This is a nice little tricky spot to be able to rely on unit test coverage.

Reply:
Maybe I'm missing something, but here's what I was thinking:

We have

```
if !propsEqual(as.props, that.props) {
  if len(as.props) == 0 {
    as.props = that.props
  }
}
```

But `if len(as.props) == 0` then either

`!propsEqual(as.props, that.props) == true`  --> do the assignment
or
`propsEqual(as.props, that.props) == false`  --> both `as.props` and `that.props` are have `len() == 0` and the assignment is a no-op, so it's okay to do.

Reply:
This is a tricky one because some of the cases it handles are technically erroneous, and thus the expected behavior is ill-defined.

One distinction to note is that `len(as.props) == 0` actually handles two cases: `as.props == nil`  and `as.props == []string{}`. Different values, but both evaluate to length 0.

|  | as.props == that.props | as.props != that.props |
| --- | --- |---|
| len(as.props) > 0  | Consistent aggregation (keep accumulating) | Conflicting aggregation (error?) |
| len(as.props) == 0 | Consistent non-aggregation (keep accumulating) | Started without props, but ended up with some; most likely scenario is that we want to aggregate by `that.props` but the window reaches too far back, so we started with some `nil` `props`. Best to adopt the new, non-nil props. |

Again, the tricky part is that aggregating by `nil` causes no aggregation, whereas aggregating by `[]string{}` causes full aggregation. They are both generate `len(props) == 0`, but they cause opposite aggregation outcomes.

My takeaways are:
- Always question logic like this, because it does look suspect. But when in doubt, it's safer to maintain existing logic.
- We need to be able to reference a unit test in this comment that enforces expacted behavior. If that unit test fails, it means we broke it. If it passes, it means that the author of the change hasn't broken anything.
- It would be justified to revisit the design that makes this a sticking point. If we "define the problem out of existence" by picking a better architecture, then we won't have to worry about this in the future. I don't know the answer here, but it's worth getting it right.